reactfire/useObservable/SuspenseSubject.ts

```diff
@@ -0,0 +1,90 @@
+import { Observable, Subject, Subscription, Subscriber, empty } from 'rxjs';
+import { tap, share, catchError } from 'rxjs/operators';
+
+export class SuspenseSubject<T> extends Subject<T> {
+  private _value: T | undefined;
+  private _hasValue = false;
+  private _timeoutHandler: NodeJS.Timeout;
+  private _innerSubscriber: Subscription;
+  private _firstEmission: Promise<void>;
+  private _resolveFirstEmission: () => void;
+  private _error: any = undefined;
+  private _innerObservable: Observable<T>;
+  private _warmupSubscription: Subscription;
+
+  constructor(innerObservable: Observable<T>, private _timeoutWindow: number) {
+    super();
+    this._firstEmission = new Promise<void>(
+      resolve => (this._resolveFirstEmission = resolve)
+    );
+    this._innerObservable = innerObservable.pipe(
+      tap(
+        v => {
+          this._next(v);
+        },
+        e => {
+          // save the error, so that we can raise on subscription or .value
+          // resolve the promise, so suspense tries again
+          this._error = e;
+          this._resolveFirstEmission();
+        }
+      ),
+      catchError(() => empty()),
+      share()
+    );
+    // warm up the observable
+    this._warmupSubscription = this._innerObservable.subscribe();
+
+    // set a timeout for reseting the cache, subscriptions will cancel the timeout
+    // and reschedule again on unsubscribe
+    this._timeoutHandler = setTimeout(this._reset, this._timeoutWindow);
+  }
+
+  get hasValue(): boolean {
+    // hasValue returns true if there's an error too
+    // so that after we resolve the promise & useObservable is called again
+    // we won't throw again
+    return this._hasValue || !!this._error;
+  }
+
+  get value(): T {
+    // TODO figure out how to reset the cache here, if I _reset() here before throwing
+    // it doesn't seem to work.
+    // As it is now, this will burn the cache entry until the timeout fires.
```

Comment:
I don't understand what this means

Reply:
Still relevant. Say for example you (end user) hit a view you didn’t have permission for and the developer wasn’t guarding against it.

You log in, navigate to the same component again within 30 seconds, and you still see the same error screen since the cache is still kickin.

Reply:
Reload the page or wait more than 30s it shows, no error.

Reply:
Gotcha, thanks!